sklearn/ensemble/_hist_gradient_boosting/_binning.pyx

```diff
@@ -70,3 +74,67 @@ cdef void _map_num_col_to_bins(const X_DTYPE_C [:] data,
                 else:
                     left = middle + 1
             binned[i] = left
+
+
+def _map_cat_to_bins(const X_DTYPE_C [:, :] data,
+                     dict bin_categories,
+                     const unsigned char missing_values_bin_idx,
+                     X_BINNED_DTYPE_C [:, :] binned):
+    """Encode categories.
+
+    Missing values and unknown values are mapped to the missing bin.
+
+    Parameters
+    ----------
+    data : ndarray of shape (n_samples, n_features)
+        data to encoded.
+    bin_categories : dict of int to arrays
+        For each categorical feature, this gives a maps categorical indicies
+        to the categories corresponding to each bin.
+    missing_values_bin_idx : uint8
+        The index of the bin where missing values are mapped.
+    binned : ndarray, shape (n_samples, n_features)
+        Output array
+    """
+    cdef:
+        long feature_idx
+        X_DTYPE_C [:] categories
+
+    for feature_idx, categories in bin_categories.items():
+        _map_cat_col_to_bins(data[:, feature_idx], categories,
+                             missing_values_bin_idx, binned[:, feature_idx])
+
+
+cdef void _map_cat_col_to_bins(const X_DTYPE_C [:] data,
+                               const X_DTYPE_C [:] categories,
+                               const unsigned char missing_values_bin_idx,
+                               X_BINNED_DTYPE_C [:] binned):
+        """Binary search for categories."""
+        cdef:
+            int i, left, right, middle
+            unsigned char found
+            X_DTYPE_C middle_value, current_value
+
+        for i in prange(data.shape[0], schedule='static', nogil=True):
+            if isnan(data[i]) or data[i] < 0:
+                binned[i] = missing_values_bin_idx
+            else:
+                current_value = data[i]
+                found = False
+                left, right = 0, categories.shape[0] - 1
+                while left <= right:
+                    middle = (left + right) // 2
```

Comment:
sort of a nit but `left + (right - left) // 2` is safe w.r.t integer overflow

I realize we don't do that above though